### ISSUE TITLE
Package import in npm

### ISSUE THREAD

**Issue by Ayc0** [stdlib-js/stdlib#187]
Currently, if I want to import `isArray` for instance, I have to either to:

```js
const stdlib = require("@stdlib/stdlib"); // 1.3M (gzipped: 344.3K)
const isArray = stdlib.assert.isArray
```

or
```js
const isArray = require("@stdlib/stdlib/lib/node_modules/@stdlib/assert/is-array"); // 9.5K (gzipped: 1K)
```

and in both cases, it is unpractical.

And if I want to import the package as written in the doc, it will fail:

```js
const isArray = require("@stdlib/assert/is-array");
// I've also tried isArray, is-array/ and isArray/
> Error: Cannot find module '@stdlib/assert/is-array'
```

My suggestion is also to publish separately every package (`@stdlib/assert` for instance) on npm (but still keep the current `@stdlib/stdlib`).

**Comment by Ayc0** [comment]
(and every folder in `lib/node_modules/@stdlib/` already has a package.json with the correct data)

**Comment by kgryte** [comment]
@Ayc0 Thanks for taking the time to file an issue. Unfortunately, while the future plan is to publish individual packages (this is why we architected the library as we have done), we cannot do so at the moment.

The rationale for this inability is that, before we publish individual packages, we need to fix (i.e., solidify) all namespaces and package names. As publishing to `npm` is permanent, once we publish a package, we cannot simply change its name in the future should we decide to rename it or move it around. And atm, we have yet to craft a coherent deprecation and publishing strategy to accommodate such changes as they will inevitably happen in the future.

In short, we have already thought about this problem, but our hands are tied until the library is further developed and solidified. It should go without saying that the sooner we are able to build out feature functionality and map future feature development, the sooner we will be able to publish everything as individual packages as you suggest.

If you would like to get involved in such an effort, that would be appreciated. :)

In any event, if you want a quick workaround if you are using stdlib in an application and not a library you plan to publish to `npm`, you can use `NODE_PATH` to resolve the `@stdlib/stdlib/lib/node_modules` folder. From there, all `@stdlib/assert/*` and every other package will be resolved as shown in the documentation.

If you are using stdlib for a frontend use case, doing `isArray = require( '@stdlib/stdlib' ).assert.isArray` is fine. With symbol resolution and "tree-shaking", your bundle should only include the symbols used by your program and not the totality of stdlib.

Lastly, while not as simple as an `npm install`, if you know what functionality you want ahead of time, you can always create a bundle (which can be done via the stdlib CLI) and vendor, similar to how one might have done back when jQuery ruled the web.

Hope the above helps!

**Comment by joshhopkins** [comment]
Being new to stdlib, I found the documentation on this to be quite unclear. Perhaps this could be elaborated further in the installation guide.

**Comment by joshhopkins** [comment]
This seems to be the simplest solution currently:

``` javascript
const stdlib = '@stdlib/stdlib/lib/node_modules/@stdlib/'
const assert = require(stdlib + 'assert')
```

**Comment by kgryte** [comment]
@joshhopkins Agreed. :( We have been actively exploring installation procedures to remove the need for this workaround when installing the entire project, but have yet to find a robust solution. Once we do find a solution, I'll be sure to update this thread.

**Comment by kgryte** [comment]
**Update**: we updated the installation process for `@stdlib/stdlib` some time ago. You can now

```bash
$ npm install @stdlib/stdlib
```

and then in JavaScript

```javascript
var isArray = require( '@stdlib/assert/is-array' );
```

We have yet to publish individual packages for individual consumption, but at least requiring individual packages now matches the documentation after installing all of `@stdlib/stdlib`.

**Comment by chriszrc** [comment]
@kgryte It would be wonderful to include this in the documentation right here:

https://github.com/stdlib-js/stdlib#installation

In the getting started section. Also, has there been any discussion of typescript typings? I can open a new issue for this if there's hasn't-

**Comment by kgryte** [comment]
@chriszrc Are you specifically referring to how to consume individual packages?

Re: TypeScript. Probably best to open a new issue. The repository includes type declarations, but you may be looking for something more specific.

**Comment by chriszrc** [comment]
Yes, specifically this:

```
var isArray = require( '@stdlib/assert/is-array' );
```

or this for typescript

```
import ttest2 from '@stdlib/stats/ttest2';
```

And wait, where are the type declarations in the repo?

**Comment by kgryte** [comment]
Re: individual packages. Noted. We'll make a TODO.

Re: TypeScript. The type declarations are present in each individual package `/docs/types` folder, which is also noted in the root [`tsconfig`](https://github.com/stdlib-js/stdlib/blob/develop/tsconfig.json) file.

**Comment by kgryte** [comment]
@chriszrc And here are the TS [docs](https://stdlib.io/docs/ts/).

**Comment by chriszrc** [comment]
Ah, I think I was thrown off by a couple of things. 

1 - I was looking at the ttest package, and it has no types:

https://github.com/stdlib-js/stdlib/tree/7a431eb78d9ea0aed1ad5559c233e5100d18625f/lib/node_modules/%40stdlib/stats/ttest/docs

Apparently all of the stats packages have no types?

2 - I did at one point find this file:

https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/types

But the index.d.ts only contains typings for a few modules? I am understanding now that adding `"typeRoots": [ "./path/to/@stdlib/types" ],` is giving me some types, but I was initially thrown off, because it didn't add any types for all of the functions I was trying out at the time (stats packages)

I suspect that for people interested in using this library with typescript, that this info should also be on the front page, or at least linked from there-

**Comment by kgryte** [comment]
@chriszrc We have TypeScript declarations for about `70%` of the packages. @Planeshifter has been working on this and adding declarations to the stats packages is on his roadmap. He may also have more insight into how you may need to set up your project to leverage the existing declarations. But this discussion would be better placed in a separate issue focused on TypeScript declarations.

**Comment by chriszrc** [comment]
I think you've answered all my questions for the moment, that's really exciting that the rest of the typescript definitions are coming, that was one of my reservations about using this library in production, very exciting and thanks for all the hard work!

**Comment by MikeMitterer** [comment]
@kgryte Really an awesome lib - to sad that you've implemented it in pure JS. Scanning through the TS-docs showed quite a few "any" types :-(

**Comment by kgryte** [comment]
@MikeMitterer Always willing to consider and accept contributions.

**Comment by kgryte** [comment]
@MikeMitterer Also not clear what you are referring to. I just did a search. Where we've used `any` has been appropriate. For example, assertion utilities are generally for testing "generics", so of course we are going to type an input argument as `any`. Further, for things like utilities which accept a `thisArg`, again, the `thisArg` can be whatever value one wants. So, I am not sure where you find the types deficient and/or problematic. Where we can be more specific, we are. If you find areas for improvement, feel free to file an issue, point to specific instances, and create a PR.
